contracts/TuiChainController.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: UNLICENSED
+/* -------------------------------------------------------------------------- */
+
+pragma solidity ^0.6.0;
+
+import "./TuiChainLoan.sol";
+import "./TuiChainMarket.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+/* -------------------------------------------------------------------------- */
+
+contract TuiChainController is Ownable
+{
+    // The Dai contract.
+    IERC20 private immutable dai;
+
+    TuiChainMarket private immutable market;
+
+    mapping(TuiChainLoan => bool) private loans;
+
+    /* ---------------------------------------------------------------------- */
+
+    /**
+     * The owner can later change the fee with setFee().
+     *
+     * @param _dai The Dai contract
+     * @param _marketFeeRecipient Address of EOA or contract to which market
+     *     fees are to be transferred
+     * @param _marketFeeAttoDaiPerNanoDai Market purchase fee, in atto-Dai per
+     *     nano-Dai
+     */
+    constructor(
+        IERC20 _dai,
+        address _marketFeeRecipient,
+        uint256 _marketFeeAttoDaiPerNanoDai
+        ) public
+    {
+        require(_dai != IERC20(0));
```

Comment:
You don't need to require that _dai address is equal to the actual dai contract address?

Reply:
The addresses of the Dai contracts in the Ethereum mainnet and the several testnets differ, so I added it as a constructor parameter instead of having it hard coded.